File: src/web_app/modern_web/src/lib/services/positioning/arrows/processors/DirectionalTupleProcessor.ts
/**
 * Directionexport interface IDirectionalTupleService {
	calculateDirectionalTuple(_motion: MotionData, location: Location): [number, number];
	generateDirectionalTuples(
		_motion: MotionData,
		baseX: number,
		baseY: number
	): Array<[number, number]>;
} Processor
 *
 * Handles complex directional tuple processing for arrow positioning adjustments.
 * Direct TypeScript port of the Python DirectionalTupleProcessor.
 *
 * This service handles:
 * - Directional tuple generation based on motion data
 * - Quadrant index calculation for proper tuple selection
 * - Complex adjustment processing for different motion types
 */

import type { MotionData } from '$lib/domain';
import { Location } from '$lib/domain';
import type { Point } from '../../interfaces';

export interface IDirectionalTupleCalculator {
	calculateDirectionalTuple(motion: MotionData, location: Location): [number, number];
	generateDirectionalTuples(
		motion: MotionData,
		baseX: number,
		baseY: number
	): Array<[number, number]>;
}

export interface IQuadrantIndexCalculator {
	calculateQuadrantIndex(location: Location): number;
}

export interface IDirectionalTupleProcessor {
	processDirectionalTuples(baseAdjustment: Point, _motion: MotionData, location: Location): Point;
}

export class DirectionalTupleCalculator implements IDirectionalTupleCalculator {
	/**
	 * Calculator for directional tuples used in arrow positioning.
	 */

	calculateDirectionalTuple(_motion: MotionData, location: Location): [number, number] {
		/**
		 * Calculate directional tuple for arrow positioning.
		 *
		 * Args:
		 *     motion: Motion data containing type and rotation direction
		 *     location: Arrow location
		 *
		 * Returns:
		 *     Tuple of [x_offset, y_offset] directional adjustments
		 */
		// This is a simplified implementation
		// In the full version, this would use complex directional calculations
		const motionType = _motion.motion_type?.toLowerCase();
		const rotationDir = _motion.prop_rot_dir?.toLowerCase();

		// Basic directional adjustments based on motion type and rotation
		let baseX = 0;
		let baseY = 0;

		if (motionType === 'pro') {
			if (rotationDir === 'clockwise' || rotationDir === 'cw') {
				baseX = 1;
				baseY = 0;
			} else {
				baseX = -1;
				baseY = 0;
			}
		} else if (motionType === 'anti') {
			if (rotationDir === 'clockwise' || rotationDir === 'cw') {
				baseX = -1;
				baseY = 0;
			} else {
				baseX = 1;
				baseY = 0;
			}
		}

		// Apply location-based adjustments
		const locationMultiplier = this.getLocationMultiplier(location);
		return [baseX * locationMultiplier.x, baseY * locationMultiplier.y];
	}

	generateDirectionalTuples(
		_motion: MotionData,
		baseX: number,
		baseY: number
	): Array<[number, number]> {
		/**
		 * Generate directional tuples for the given motion and base adjustment.
		 *
		 * Args:
		 *     motion: Motion data containing type, rotation, and location info
		 *     baseX: Base X adjustment value
		 *     baseY: Base Y adjustment value
		 *
		 * Returns:
		 *     Array of 4 directional tuples representing rotated adjustments
		 */
		const tuples: Array<[number, number]> = [];

		// Generate 4 rotational variants of the base adjustment
		// This implements the 4-tuple rotation logic from the reference
		for (let i = 0; i < 4; i++) {
			const angle = (i * 90 * Math.PI) / 180; // Convert to radians
			const cos = Math.cos(angle);
			const sin = Math.sin(angle);

			// Apply rotation matrix
			const rotatedX = baseX * cos - baseY * sin;
			const rotatedY = baseX * sin + baseY * cos;

			tuples.push([Math.round(rotatedX), Math.round(rotatedY)]);
		}

		return tuples;
	}

	private getLocationMultiplier(location: Location): { x: number; y: number } {
		/**Get location-based multiplier for directional adjustments.*/
		const multipliers: Record<Location, { x: number; y: number }> = {
			[Location.NORTH]: { x: 0, y: -1 },
			[Location.NORTHEAST]: { x: 1, y: -1 },
			[Location.EAST]: { x: 1, y: 0 },
			[Location.SOUTHEAST]: { x: 1, y: 1 },
			[Location.SOUTH]: { x: 0, y: 1 },
			[Location.SOUTHWEST]: { x: -1, y: 1 },
			[Location.WEST]: { x: -1, y: 0 },
			[Location.NORTHWEST]: { x: -1, y: -1 },
		};

		return multipliers[location] || { x: 0, y: 0 };
	}
}

export class QuadrantIndexCalculator implements IQuadrantIndexCalculator {
	/**
	 * Calculator for quadrant indices used in directional tuple selection.
	 */

	calculateQuadrantIndex(location: Location): number {
		/**
		 * Calculate quadrant index for the given location.
		 *
		 * Args:
		 *     location: Arrow location
		 *
		 * Returns:
		 *     Quadrant index (0-3)
		 */
		const quadrantMap: Record<Location, number> = {
			[Location.NORTHEAST]: 0,
			[Location.SOUTHEAST]: 1,
			[Location.SOUTHWEST]: 2,
			[Location.NORTHWEST]: 3,
			// Cardinal directions map to nearest quadrant
			[Location.NORTH]: 0, // Maps to NE quadrant
			[Location.EAST]: 1, // Maps to SE quadrant
			[Location.SOUTH]: 2, // Maps to SW quadrant
			[Location.WEST]: 3, // Maps to NW quadrant
		};

		return quadrantMap[location] || 0;
	}
}

export class DirectionalTupleProcessor implements IDirectionalTupleProcessor {
	/**
	 * Processor for applying directional tuple adjustments to arrow positioning.
	 */

	constructor(
		private directionalTupleService: IDirectionalTupleCalculator,
		private quadrantIndexService: IQuadrantIndexCalculator
	) {}

	processDirectionalTuples(
		baseAdjustment: Point,
		_motion: MotionData,
		location: Location
	): Point {
		/**
		 * Process directional tuples to calculate final adjustment.
		 *
		 * Args:
		 *     baseAdjustment: Base adjustment from placement services
		 *     motion: Motion data for directional calculations
		 *     location: Arrow location for quadrant selection
		 *
		 * Returns:
		 *     Final adjustment point after directional processing
		 */
		try {
			// Generate directional tuples from base adjustment
			const directionalTuples = this.directionalTupleService.generateDirectionalTuples(
				_motion,
				baseAdjustment.x,
				baseAdjustment.y
			);

			// Calculate quadrant index for tuple selection
			const quadrantIndex = this.quadrantIndexService.calculateQuadrantIndex(location);

			// Select the appropriate tuple based on quadrant
			const selectedTuple = directionalTuples[quadrantIndex] || [0, 0];

			// Calculate directional tuple for additional adjustments
			const directionalTuple = this.directionalTupleService.calculateDirectionalTuple(
				_motion,
				location
			);

			// Combine base adjustment, selected tuple, and directional tuple
			const finalX = baseAdjustment.x + selectedTuple[0] + directionalTuple[0];
			const finalY = baseAdjustment.y + selectedTuple[1] + directionalTuple[1];

			return { x: finalX, y: finalY };
		} catch (error) {
			console.warn('Directional tuple processing failed, using base adjustment:', error);
			return baseAdjustment;
		}
	}
}

File: src/desktop/shared/application/services/positioning/arrows/orchestration/directional_tuple_processor.py
"""
Directional Tuple Processor Service

Focused service for handling directional tuple generation and quadrant-based selection.
Extracted from the god class ArrowAdjustmentCalculatorService.

RESPONSIBILITIES:
- Generate directional tuples using rotation matrices
- Calculate quadrant indices for selection
- Select final adjustment from tuples
- Error handling with Result types

USAGE:
    processor = DirectionalTupleProcessor(
        directional_tuple_service,
        quadrant_index_service
    )

    try:
        final_adjustment = processor.process_directional_tuples(base_adjustment, motion_data, location)
    except Exception as e:
        logger.error(f"Processing failed: {e}")
"""

import logging

from desktop.modern.core.types.coordinates import point_to_qpoint
from desktop.modern.core.types.geometry import Point
from desktop.modern.domain.models.enums import Location
from desktop.modern.domain.models.motion_data import MotionData

# Import required services
from ...arrows.calculation.directional_tuple_calculator import (
    DirectionalTupleCalculator,
)
from ..calculation.quadrant_index_calculator import QuadrantIndexCalculator

logger = logging.getLogger(__name__)


class DirectionalTupleProcessor:
    """
    Focused service for directional tuple processing.

    Handles the tuple generation and selection phase of arrow positioning:
    1. Generate directional tuples using rotation matrices
    2. Calculate quadrant index for selection
    3. Select final adjustment from tuples
    4. Return proper Result types with error handling
    """

    def __init__(
        self,
        directional_tuple_service: DirectionalTupleCalculator,
        quadrant_index_service: QuadrantIndexCalculator,
    ):
        """Initialize with required services for tuple processing."""
        self.directional_tuple_service = directional_tuple_service
        self.quadrant_index_service = quadrant_index_service

    def process_directional_tuples(
        self,
        base_adjustment: Point,
        motion_data: MotionData,
        location: Location,
    ) -> Point:
        """
        Process directional tuples to get final adjustment.

        Args:
            base_adjustment: Base adjustment point from lookup
            motion_data: Motion data containing type, rotation, and location info
            location: Pre-calculated arrow location
        """

        try:
            # STEP 1: Generate directional tuples using base adjustment
            directional_tuples = self._generate_directional_tuples(
                motion_data, base_adjustment
            )

            # STEP 2: Get quadrant index for selection using pre-calculated location
            quadrant_index = self._get_quadrant_index(motion_data, location)

            # STEP 3: Select final adjustment from directional tuples
            final_adjustment = self._select_from_tuples(
                directional_tuples, quadrant_index
            )

            return final_adjustment

        except Exception as e:
            logger.error(f"Error processing directional tuples: {e}")
            raise RuntimeError(f"Directional tuple processing failed: {e}") from e

    def _generate_directional_tuples(
        self, motion: MotionData, base_adjustment: Point
    ) -> list[tuple[int, int]]:
        """
        Generate directional tuples using rotation matrices.

        Returns Result[List[Tuple[int, int]], AppError] instead of list with fallback.
        """
        try:
            # Convert Point to QPointF for legacy service compatibility
            base_qpoint = point_to_qpoint(base_adjustment)

            directional_tuples = (
                self.directional_tuple_service.generate_directional_tuples(
                    motion, base_qpoint.x(), base_qpoint.y()
                )
            )

            if not directional_tuples:
                raise ValueError("Directional tuple service returned empty tuples")

            return directional_tuples

        except Exception as e:
            logger.error(f"Error generating directional tuples: {e}")
            raise RuntimeError(f"Directional tuple generation failed: {e}") from e

    def _get_quadrant_index(self, motion_data: MotionData, location: Location) -> int:
        """
        Get quadrant index for directional tuple selection using pre-calculated location.
        """
        try:
            # Use pre-calculated location instead of recalculating
            quadrant_index = self.quadrant_index_service.get_quadrant_index(
                motion_data, location
            )

            # Validate quadrant index
            if quadrant_index < 0:
                raise ValueError(f"Invalid negative quadrant index: {quadrant_index}")

            return quadrant_index

        except Exception as e:
            logger.error(f"Error calculating quadrant index: {e}")
            raise RuntimeError(f"Quadrant index calculation failed: {e}") from e

    def _select_from_tuples(
        self, directional_tuples: list[tuple[int, int]], quadrant_index: int
    ) -> Point:
        """
        Select final adjustment from directional tuples using quadrant index.
        """
        try:
            if not directional_tuples:
                raise ValueError("Cannot select from empty directional tuples")

            if quadrant_index >= len(directional_tuples):
                raise ValueError(
                    f"Quadrant index {quadrant_index} out of range for {len(directional_tuples)} tuples"
                )

            selected_tuple = directional_tuples[quadrant_index]
            final_point = Point(float(selected_tuple[0]), float(selected_tuple[1]))

            return final_point

        except Exception as e:
            logger.error(f"Error selecting from directional tuples: {e}")
            raise RuntimeError(f"Tuple selection failed: {e}") from e
